### PR TITLE
Prototype grpc server

### PR DESCRIPTION
**Pull request by dsyer**
The app now listens on port 10382 for a stream of riff
Messages, processing them using a `Function` from
the URI passed in on startup.

### CODE REVIEW

**Comment by markfisher** [comment]
getting build failures with this in the log:

```
java.lang.ClassNotFoundException: io.netty.buffer.PoolArena$1
```

these are the errors:

```
Tests in error:
  FunctionConfigurationTests$ConsumerCompositionTests.testConsumer » IllegalState
  FunctionConfigurationTests$FunctionCompositionTests.testFunction » IllegalState
  FunctionConfigurationTests$FunctionCompositionTests.testThen » IllegalState Fa...
  FunctionConfigurationTests$SingleFunctionTests.testDouble » IllegalState Faile...
  FunctionConfigurationTests$SupplierCompositionTests.testFunction » IllegalState
  FunctionConfigurationTests$SupplierCompositionTests.testSupplier » IllegalState

Tests run: 18, Failures: 0, Errors: 6, Skipped: 0
```

**Comment by dsyer** [comment]
Bizarre error (cause still unknown - something to do with an abstract test class). I think it's fixed now. Not sure what the abstract class does.

**Comment by markfisher** [comment]
yes, the build passes now

thanks!

File: src/main/java/io/projectriff/invoker/JavaFunctionInvokerServer.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+			@Override
+			public void onCompleted() {
+				emitter.onComplete();
+			}
+		};
+
+	}
+
+	private Message<byte[]> payloadToBytes(Message<?> message) {
```
Comment (dturanski) [inline src/main/java/io/projectriff/invoker/JavaFunctionInvokerServer.java:96]: Maybe extract from here down, along with the mapper, to a separate class

File: src/main/java/io/projectriff/invoker/MessageConversionUtils.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+	public static Message<byte[]> fromGrpc(
+			io.projectriff.grpc.function.FunctionProtos.Message input) {
+		MessageBuilder<byte[]> builder = MessageBuilder
+				.withPayload(input.getPayload().toByteArray());
+		for (Entry<String, HeaderValue> entry : input.getHeadersMap().entrySet()) {
+			HeaderValue header = entry.getValue();
+			if (header.getValuesCount() > 0) {
+				Object value;
+				ProtocolStringList list = header.getValuesList();
```
Comment (dturanski) [inline src/main/java/io/projectriff/invoker/MessageConversionUtils.java:45]: ProtocolStringList can be a List here

File: src/test/java/io/projectriff/invoker/GrpcTestClient.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * @author Dave Syer
+ *
+ */
+public class GrpcTestClient {
+	private final Log logger = LogFactory.getLog(GrpcTestClient.class);
+
+	private final ManagedChannel channel;
+	private final MessageFunctionStub asyncStub;
+
+	/** Construct client for accessing RouteGuide server at {@code host:port}. */
```
Comment (ericbottard) [inline src/test/java/io/projectriff/invoker/GrpcTestClient.java]: Comment leftover from gRPC getting started guide I believe. will remove

**Comment by ericbottard** [comment]
Can't make this to work with the `Greeter` sample from riif, modified to return headers like this (needed to that `correlationId` is propagated):

```
public class Greeter implements Function<Message<String>, Message<String>> {

	public Message<String> apply(Message<String> in) {
		return MessageBuilder.withPayload("Hello " + in.getPayload())
				.setHeaders(new MessageHeaderAccessor(in)).build();
	}
}
```

The `Message` class is not found.

Tried with `--function.runner.isolated=false`, or with a boot packaged uberjar. No luck.

**Comment by ericbottard** [comment]
I have a rebased branch with this, with some additional polish that you may want to grab at https://github.com/ericbottard/java-function-invoker/tree/grpc

**Comment by dsyer** [comment]
@ericbottard I don't think that message function would work on master either. Did you say you had it working there (confused)?

**Comment by ericbottard** [comment]
No, didn't say (or looked) that it worked on master either.

This is prompted by the fact that, switching to this gRPC style, we now need to take care of propagating the `correlationId` at the function level (at least for now), so this PR would be useless if we can't use `spring-messaging` in the function classloader

**Comment by dsyer** [comment]
It doesn't work on master then, as I suspected. We can discuss OOB.

**Comment by jldec** [comment]
from OOB discussion today
> i don't have enough context to offer guidance on the `<Message>` idea
> i do think it would be ok to release 0.0.4 with a java invoker that supports req/resp on http and flux streams (without correlation) on grpc if that is easier

**Comment by dsyer** [comment]
Superseded by #26.

**Pull request closed without merging**